PyDSS/pyPostprocessor/PostprocessScripts/postprocess_voltage_upgrades.py

```diff
@@ -19,390 +19,160 @@
+        self.orig_reg_controls = self.Settings["orig_reg_controls"]
+        self.new_capacitors = self.Settings["new_capacitors"]
+        self.orig_capacitors = self.Settings["orig_capacitors"]
+        self.new_capcontrols = self.Settings["new_capcontrols"]
+        self.orig_capcontrols = self.Settings["orig_capcontrols"]
+        self.orig_ckt_info = self.Settings["orig_ckt_info"]
+        self.orig_xfmr_info = self.Settings["orig_xfmr_info"]
+        self.final_cap_upgrades = {}
+        self.final_reg_upgrades = {}
+        self.processed_outputs = {}
 
-        self.cap_processed_ops["feederhead_name"] = self.Settings["feederhead_name"]
-        self.cap_processed_ops["feederhead_basekV"] = self.Settings["feederhead_basekV"]
+        # convert input list of nested dictionaries to a dictionary of nested dictionaries
+        self.new_reg_controls = {k: v for element in self.new_reg_controls for k, v in element.items()}
+        self.orig_reg_controls = {k: v for element in self.orig_reg_controls for k, v in element.items()}
+        self.new_capacitors = {k: v for element in self.new_capacitors for k, v in element.items()}
+        self.orig_capacitors = {k: v for element in self.orig_capacitors for k, v in element.items()}
+        self.new_capcontrols = {k: v for element in self.new_capcontrols for k, v in element.items()}
+        self.orig_capcontrols = {k: v for element in self.orig_capcontrols for k, v in element.items()}
 
-        self.write_to_json(self.cap_processed_ops,"Processed_voltage_upgrades")
+        if self.new_capcontrols:
+            self.get_capacitor_upgrades()
+        if self.new_reg_controls:
+            self.get_regulator_upgrades()
 
-    def write_to_json(self, dict, file_name):
-        with open(os.path.join(self.Settings["outputs"],"{}.json".format(file_name)), "w") as fp:
-            json.dump(dict, fp, indent=4)
+        self.processed_outputs["feederhead_name"] = self.Settings["feederhead_name"]
+        self.processed_outputs["feederhead_basekV"] = self.Settings["feederhead_basekV"]
+        self.write_to_json(self.processed_outputs, "Processed_voltage_upgrades")
 
-    def read_orig_upgrades(self):
-        f = open(os.path.join(self.Settings["outputs"],"Initial_capacitors.json"),"r")
-        self.orig_caps = json.load(f)
-        f = open(os.path.join(self.Settings["outputs"], "Initial_regulators.json"), "r")
-        self.orig_regs = json.load(f)
+    # function to compare two dictionaries with same format
+    # only compares common elements present in both original and new dictionaries
+    def compare_dict(self, old, new):
+        field_list = []
+        change = {}
+        sharedKeys = set(old.keys()).intersection(new.keys())
+        for key in sharedKeys:
+            change_flag = False
+            for sub_field in old[key]:
+                if old[key][sub_field] != new[key][sub_field]:
+                    change_flag = True
+                    field_list.append(sub_field)
```

Comment:
@daniel-thom: we might want to make use of this function for thermal upgrades as well. Is there a better place to keep this, or let it stay in this file?

Reply:
You could make a new file in PyDSS/pyPostprocessor.  You could also put the function in PyDSS/utils/utils.py.

Reply:
Okay, can do that when doing thermal.